pkgs/development/go-modules/generic/default.nix

```diff
@@ -53,7 +56,46 @@ let
 
   vendCommand = if runVend then "${vend}/bin/vend" else "false";
 
-  go-modules = if vendorSha256 != null then go.stdenv.mkDerivation (let modArgs = {
+  dep2src = goDep:
+    {
+      inherit (goDep) goPackagePath;
+      src = if goDep.fetch.type == "git" then
+        fetchgit {
+          inherit (goDep.fetch) url rev sha256;
+        }
+      else if goDep.fetch.type == "hg" then
+        fetchhg {
+          inherit (goDep.fetch) url rev sha256;
+        }
+      else if goDep.fetch.type == "bzr" then
+        fetchbzr {
+          inherit (goDep.fetch) url rev sha256;
+        }
+      else if goDep.fetch.type == "FromGitHub" then
+        fetchFromGitHub {
+          inherit (goDep.fetch) owner repo rev sha256;
+        }
+      else abort "Unrecognized package fetch type: ${goDep.fetch.type}";
+    };
```

Comment:
I was wondering we could not just accept file that is called with `callPackage`.
The file would have all required fetchers in its header:

```
{ fetchgit, fetchhg }: {
   "github.com/foo" = fetchgit { ... };
   "somehost.com/foo" = fetchhg {};
}
```

Advantage is better extensibility. And we could replace/patch dependencies by its go path.
The patching could be even done with a global override file so it does not need to be done per package.